Posts/2023/06/Measuring-Download-Time.md

```diff
@@ -0,0 +1,544 @@
+---
+post_title: 'Measuring average download time'
+username: francisconabas
+categories: PowerShell
+tags: PowerShell, Automation, Performance, Measure-Command
+summary: This post shows how to measure average download time with PowerShell
+---
+
+One of the most overlooked roles of a systems administrator is to be able to troubleshoot
+network issues. How many times had you been in a situation where your servers are problematic,
+and someone asked you to check the network connectivity?
+One of the steps is checking downloading time, and speed, and although there are countless
+tools available, today we will learn how to do it natively, with PowerShell.
+
+## Methods
+
+We will focus on three methods, ranging from the easiest to the most complex, and discuss
+their pros and cons. These methods are the `Start-BitsTransfer` Cmdlet, using .NET with the
+`System.Net` namespace, and using the Windows native API.
+
+### `Start-BitsTransfer`
+
+BITS, or Background Intelligent Transfer Service is a Windows service that manages content transfer
+using HTTP or SMB. It was designed to manage the many aspects of content transfer, including cost,
+speed, priority, etc.
+For us, it also serves as an easy way of downloading files. Here is how you download a file from a
+web server using BITS:
+
+```powershell
+Start-BitsTransfer -Source 'https://www.myawesomeserver.com/Files/BitsDefinition.txt' `
+-Destination 'C:\BitsDefinition.txt'
```

Comment:
Don't use backtick for line continuation. Use splatting instead.
```suggestion
$startBitsTransferSplat = @{
    Source = 'https://www.contoso.com/Files/BitsDefinition.txt'
    Destination = 'C:\BitsDefinition.txt'
}
Start-BitsTransfer @startBitsTransferSplat
```

Also, I changed the hostname to `www.contoso.com`. That is a reserved name that Microsoft owns. It is guaranteed to not be used by other people. While `myawesomeserver.com` is not currently registered, a malicious person could register it and point to malware, etc.